// @ui-kitten/components/ui/select/selectOptionsList.component.js
/**
 * @license
 * Copyright Akveo. All Rights Reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 */
import React from 'react';
import { List, } from '../list/list.component';
import { SelectOption, } from './selectOption.component';
import { SelectGroupOption, } from './selectGroupOption.component';
export class SelectOptionsList extends React.Component {
    constructor() {
        super(...arguments);
        this.onSelect = (option, event) => {
            this.props.onSelect(option, event);
        };
        this.hasItems = (item) => {
            return item.items && item.items.length !== 0;
        };
        this.renderGroupItem = (info) => {
            const { renderItem, multiSelect, strategy } = this.props;
            return (React.createElement(SelectGroupOption, Object.assign({}, info, { strategy: strategy, multi: multiSelect, renderItem: renderItem, onPress: this.onSelect })));
        };
        this.renderSingleItem = (info) => {
            return (React.createElement(SelectOption, Object.assign({}, info, { disabled: info.item.disabled, selected: this.props.strategy.isSelected(info.item), multi: this.props.multiSelect, onPress: this.onSelect })));
        };
        this.renderItem = (info) => {
            return this.hasItems(info.item) ? this.renderGroupItem(info) : this.renderSingleItem(info);
        };
    }
    render() {
        return (React.createElement(List, Object.assign({}, this.props, { renderItem: this.renderItem })));
    }
}
